src/server/scene/basic_idle_hub.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2021 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored By: William Wold <william.wold@canonical.com>
+ */
+
+#include "basic_idle_hub.h"
+#include "mir/time/alarm.h"
+#include "mir/time/alarm_factory.h"
+
+namespace ms = mir::scene;
+namespace mt = mir::time;
+
+namespace
+{
+auto add_awake_at_start(
+    std::vector<ms::BasicIdleHub::StateEntry> const& state_sequence)-> std::vector<ms::BasicIdleHub::StateEntry>
+{
+    std::vector<ms::BasicIdleHub::StateEntry> result{
+        {std::chrono::milliseconds{0}, ms::IdleState::awake}};
+    result.insert(result.end(), state_sequence.begin(), state_sequence.end());
+    return result;
+}
+}
+
+ms::BasicIdleHub::BasicIdleHub(
+        std::vector<StateEntry> const& state_sequence,
+        std::shared_ptr<mt::AlarmFactory> const& alarm_factory)
+    : state_sequence{add_awake_at_start(state_sequence)},
+      next_state_alarm{alarm_factory->create_alarm([this]()
+          {
+              increment_state();
+          })},
+      current_state_index{0}
+{
+    std::unique_lock<std::mutex> lock{mutex};
```

Comment:
There's no need to lock in the constructor; it's not possible for this object to be shared between threads yet.

Reply:
This is to call `set_state_index()` on the next line, which takes and unlocks a lock. It could be avoided (ie by using an optional lock) if there was any reason to, but since it's guaranteed to be uncontested and thus pretty fast this is simpler.